cmd/akamai-purger/main.go

```diff
@@ -191,38 +196,64 @@ func (ap *akamaiPurger) len() int {
 	return len(ap.toPurge)
 }
 
-func (ap *akamaiPurger) purge() error {
-	ap.Lock()
-	queueEntries := ap.toPurge[:]
-	ap.toPurge = [][]string{}
-	ap.Unlock()
-	if len(queueEntries) == 0 {
-		return nil
+func (ap *akamaiPurger) purgeBatch(batch [][]string) error {
+	// Flatten the batch of stack entries into a single slice of URLs.
+	var urls []string
+	for _, url := range batch {
+		urls = append(urls, url...)
 	}
 
-	stoppedAt, err := ap.client.Purge(queueEntries)
+	err := ap.client.Purge(urls)
 	if err != nil {
 		ap.Lock()
-
-		// Add the remaining queue entries back, but at the end of the queue. If somehow
-		// there's a URL which repeatedly results in error, it won't block the
-		// entire queue, only a single batch.
-		ap.toPurge = append(ap.toPurge, queueEntries[stoppedAt:]...)
-		ap.Unlock()
-		ap.log.Errf("Failed to purge OCSP responses for %d certificates: %s", len(queueEntries), err)
+		defer ap.Unlock()
+		if len(ap.toPurge) >= ap.maxStackSize {
+			// Drop the oldest entry from the bottom of the stack to make room.
+			ap.toPurge = ap.toPurge[len(batch):]
+		}
+		// Add the entries from the failed batch to the bottom of the stack to
+		// avoid blocking.
+		ap.toPurge = append(batch, ap.toPurge...)
```

Comment:
If we go this route - it's inefficient to shorten the slice by chopping N items off the front, then allocate a new slice just so you can prepend exactly N items. Instead you could overwrite those N items with the N new ones.

But coming back to the question of what to do about errored batches: We already know we have to drop some things on the floor when we hit our limits. Should we drop the thing at the bottom of the stack (which is somewhat old but hasn't had any tries yet) or should we drop the thing we recently pulled off the top of the stack, which had a series of failures? It will be newer (currently), so maybe? But on the whole I think it's six of one, half dozen of the other, and dropping the errored batch rather than reinserting it at the bottom of the stack is much simpler conceptually.

Reply:
> If we go this route - it's inefficient to shorten the slice by chopping N items off the front, then allocate a new slice just so you can prepend exactly N items. Instead you could overwrite those N items with the N new ones.

Ah, yeah, I can see that.

> But coming back to the question of what to do about errored batches: We already know we have to drop some things on the floor when we hit our limits. Should we drop the thing at the bottom of the stack (which is somewhat old but hasn't had any tries yet) or should we drop the thing we recently pulled off the top of the stack, which had a series of failures? It will be newer (currently), so maybe? But on the whole I think it's six of one, half dozen of the other, and dropping the errored batch rather than reinserting it at the bottom of the stack is much simpler conceptually.

Okay, I can get behind that.